### DISCUSSION TITLE
Wannier Projected Density of States

### DISCUSSION THREAD

**Discussion by ShinjanM** [wannier-berri/wannier-berri#254] [General]
Is there any provision to compute the Wannier projected density of states similar to the one available in postw90?

**Comment by stepan-tsirkin** [comment]
Dear @ShinjanM , 
your message came during the summer holidays, and was lost. Sorry for that.

Actually, it is quite straightforward to make projected DOS. First one needs to define the "Formula" for projection, which may be a class like this:

``` python
from wannierberri.formula import Matrix_ln
from wannierberri.symmetry import transform_ident

class FormulaProjWF(Matrix_ln):

    def __init__(self, data_K, wf_list):
        proj_matrix = np.zeros( (data_K.num_wann, data_K.num_wann) )
        for wf in wf_list:
            proj_matrix[wf,wf]=1
        projected = np.array( [U.T.conj().dot(proj_matrix).dot(U) for U in data_K.UU_K])
        super().__init__(  projected )
        self.ndim = 0
        self.transformTR = transform_ident
        self.transformInv = transform_ident
```

Next, if you want to plot fatbands based on weight on WFs, you can define your tabulators as

```     python
tabulators = { 
                "proj_up" : wberri.calculators.tabulate.Tabulator(FormulaProjWF, kwargs_formula = {"wf_list":np.arange(0,18,2)}),
                "proj_down" : wberri.calculators.tabulate.Tabulator(FormulaProjWF, kwargs_formula = {"wf_list":np.arange(1,18,2)}),
                "proj_s" : wberri.calculators.tabulate.Tabulator(FormulaProjWF, kwargs_formula = {"wf_list":np.arange(0,2)}),
                "proj_p" : wberri.calculators.tabulate.Tabulator(FormulaProjWF, kwargs_formula = {"wf_list":np.arange(2,8)}),
                "proj_d" : wberri.calculators.tabulate.Tabulator(FormulaProjWF, kwargs_formula = {"wf_list":np.arange(8,18)}),
                "spin": wberri.calculators.tabulate.Spin()
             }
```

Looks like the results make sense: 

<img src="https://github.com/wannier-berri/wannier-berri/assets/22172977/37caf824-1b6c-4760-84f6-b9e670b61d9d" width=200>
<img src="https://github.com/wannier-berri/wannier-berri/assets/22172977/226ae976-3940-409a-8afc-5dd425ccbcd6" width=200>
<img src="https://github.com/wannier-berri/wannier-berri/assets/22172977/27340ec1-9beb-44ed-bb8e-4e4bbfb6268c" width=200>

Further, if you want an integrated PDOS over k-space, you define the calculator as 

``` python
from wannierberri.calculators.static import DOS
class PDOS(DOS):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.Formula = FormulaProjWF
        
calculators["proj_up"] = PDOS(Efermi=Efermi, tetra=True, kwargs_formula = {"wf_list":np.arange(0,18,2)})
calculators["proj_down"] = PDOS(Efermi=Efermi, tetra=True, kwargs_formula = {"wf_list":np.arange(1,18,2)})
calculators ["dos"] = wberri.calculators.static.DOS(Efermi=Efermi, tetra=True,)
```

Here I attach the zipped notebook illustrating this : 
[tutorial-wb-basic-projected.zip](https://github.com/wannier-berri/wannier-berri/files/13660910/tutorial-wb-basic-projected.zip)
( it is derived from the [basic tutorial ](https://github.com/wannier-berri/WannierBerri-tutorial/blob/main/tutorials/1_basic/tutorial-wb-basic.ipynb) )
